File: app/src/main/java/com/example/covidnow/services/LocationUpdatesIntentService.java
package com.example.covidnow.services;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.util.Log;

import com.example.covidnow.helpers.Utils;
import com.google.android.gms.location.LocationResult;

import java.util.List;

import static com.example.covidnow.helpers.Utils.*;

/**
 * Handles incoming location updates and displays a notification with the location data.
 *
 * For apps targeting API level 25 ("Nougat") or lower, location updates may be requested
 * using {@link android.app.PendingIntent#getService(Context, int, Intent, int)} or
 * {@link android.app.PendingIntent#getBroadcast(Context, int, Intent, int)}. For apps targeting
 * API level O, only {@code getBroadcast} should be used.
 *
 *  Note: Apps running on "O" devices (regardless of targetSdkVersion) may receive updates
 *  less frequently than the interval specified in the
 *  {@link com.google.android.gms.location.LocationRequest} when the app is no longer in the
 *  foreground.
 */
public class LocationUpdatesIntentService extends IntentService {

    private static final String ACTION_PROCESS_UPDATES =
            "com.google.android.gms.location.sample.locationupdatespendingintent.action" +
                    ".PROCESS_UPDATES";
    private static final String TAG = LocationUpdatesIntentService.class.getSimpleName();


    public LocationUpdatesIntentService() {
        // Name the worker thread.
        super(TAG);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        Log.i(TAG, "Handling intent");
        if (intent != null) {
            final String action = intent.getAction();
            if (ACTION_PROCESS_UPDATES.equals(action)) {
                LocationResult result = LocationResult.extractResult(intent);
                if (result != null) {
                    List<Location> locations = result.getLocations();
                    setLocationUpdatesResult(this, locations);
                    sendNotification(this, getLocationResultTitle(this, locations));
                    Log.i(TAG, getLocationUpdatesResult(this));
                }
            }
        }
    }
}

File: app/src/main/java/com/example/covidnow/services/LocationUpdatesIntentService.kt
package com.example.covidnow.services

import android.app.IntentService
import android.content.Intent
import android.util.Log
import com.example.covidnow.helpers.Utils.getLocationUpdatesResult
import com.example.covidnow.helpers.Utils.sendNotification
import com.example.covidnow.helpers.Utils.setLocationUpdatesResult
import com.google.android.gms.location.LocationResult

/**
 * Handles incoming location updates and displays a notification with the location data.
 */
class LocationUpdatesIntentService : IntentService(TAG) {
    override fun onHandleIntent(intent: Intent?) {
        Log.i(TAG, "Handling intent")
        if (intent != null) {
            val action = intent.action
            if (ACTION_PROCESS_UPDATES == action) {
                val result = LocationResult.extractResult(intent)
                if (result != null) {
                    val locations = result.locations
                    setLocationUpdatesResult(this, locations)
                    // TODO
                    sendNotification(this, "Handling current location")
                    getLocationUpdatesResult(this)?.let { Log.i(TAG, it) }
                }
            }
        }
    }

    companion object {
        private const val ACTION_PROCESS_UPDATES = "com.google.android.gms.location.sample.locationupdatespendingintent.action" +
                ".PROCESS_UPDATES"
        private val TAG = LocationUpdatesIntentService::class.java.simpleName
    }
}